### PR TITLE
feat: signal binding for element property (#22408)

### LINKED ISSUE #22408: Add signal binding to property in Element
### Describe your motivation

`Element::bindProperty(String, Signal)` method that establishes a binding between signal instance and element/component: element's property is updated on UI automatically whenever the signal value changes and signal value is updated whenever the attribute is changed on the client side.

Flow supports `boolean`, `String`, `double`, `Object`, JSON, `Map`, `List` types for properties, thus this binding method should support the same types.

When `addPropertyChangeListener` is used, it should be invoked when property is changed and signal binding is established, but only when element is attached.

Binding a bean will be simplified after https://github.com/vaadin/flow/issues/8218#issuecomment-3351145076 is done for `Element`.

`Element::bindProperty(String, Signal)` should satisfy the general criteria given in the parent ticket.

### PR DESCRIPTION
**Pull request by tepi**
Adds signal binding for Element properties

Fixes #22408

### CODE REVIEW

File: flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java [hunk 1]
```diff
@@ -112,6 +117,75 @@ public void setProperty(String name, Serializable value) {
...
+
+        if (value == null) {
+            setProperty(name, JacksonUtils.nullNode(), emitChange);
+        } else {
+            if (value instanceof String || value instanceof Number
+                    || value instanceof Boolean
+                    || value instanceof BaseJsonNode) {
+                setProperty(name, value, emitChange);
+            } else if (value instanceof List) {
+                setProperty(name, JacksonUtils.listToJson((List<?>) value),
```
Comment (tltv) [inline flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java]: if blocks for List and Map looks unnecessary since `JacksonUtils.beanToJson` is identical implementation with `JacksonUtils.listToJson` and `JacksonUtils.mapToJson`.

Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java]: Good point, I'll simplify it. Makes me wonder though why we have the separate methods then, maybe due some historical reason?

Comment (tltv) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java]: It used [JsonUtils](https://github.com/vaadin/flow/blob/24.9/flow-server/src/main/java/com/vaadin/flow/internal/JsonUtils.java) before, which had some difference. Now it's just backwards compatible API if nothing else.

File: flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java [hunk 2]
```diff
@@ -112,6 +117,66 @@ public void setProperty(String name, Serializable value) {
...
+     * @param signal
+     *            the signal to bind or <code>null</code> to unbind any existing
+     *            binding
+     * @throws com.vaadin.signals.BindingActiveException
+     *             thrown when there is already an existing binding for the
+     *             given property
+     */
+    public void bindSignal(Element owner, String name, Signal<?> signal) {
+        SignalBinding previousSignalBinding;
+        if (super.getProperty(name) instanceof SignalBinding binding) {
```
Comment (tltv) [inline flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java]: This should call `super.get(name)`, otherwise it's always returning actual value and `previousSignalBinding` is always null.

File: flow-server/src/test/java/com/vaadin/flow/dom/ElementBindPropertyTest.java [hunk 3]
```diff
@@ -0,0 +1,864 @@
...
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.mockStatic;
+import static org.mockito.Mockito.when;
+
+public class ElementBindPropertyTest {
```
Comment (tltv) [inline flow-server/src/test/java/com/vaadin/flow/dom/ElementBindPropertyTest.java:67]: Add a test for `bindProperty("foo", null)` to verify that existing binding is removed.

File: flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java (hunk 2 above)
Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java]: Good catch, thanks! Moved the new/overridden get method from `AbstractPropertyMap` to `ElementPropertyMap` to fix this.

File: flow-server/src/test/java/com/vaadin/flow/dom/ElementBindPropertyTest.java (hunk 3 above)
Comment (tepi) [reply flow-server/src/test/java/com/vaadin/flow/dom/ElementBindPropertyTest.java:67]: Added test

File: flow-server/src/main/java/com/vaadin/flow/dom/Element.java [hunk 4]
```diff
@@ -848,6 +848,53 @@ public Element setPropertyMap(String name, Map<String, ?> value) {
...
+     * While a Signal is bound to a property, any attempt to set the property
+     * value manually throws {@link BindingActiveException}. Same happens when
+     * trying to bind a new Signal while one is already bound.
+     * <p>
+     * Supported data types for the signal are the same as for the various
+     * {@code setProperty} methods in this class: {@link String},
+     * {@link Boolean}, {@link Double}, {@link BaseJsonNode}, {@link Object}
+     * (bean), {@link List} and {@link Map}. Typed Lists and Maps are not
+     * supported, i.e. the signal must be of type {@code Signal<List<?>>} or
+     * {@code Signal<Map<String,?>}.
```
Comment (tltv) [inline flow-server/src/main/java/com/vaadin/flow/dom/Element.java]: Shouldn't it say `Signal<Map<?,?>>`. Having String type for key looks a bit misleading.

File: flow-server/src/main/java/com/vaadin/flow/dom/Element.java [hunk 5]
```diff
@@ -848,6 +848,53 @@ public Element setPropertyMap(String name, Map<String, ?> value) {
...
+     *
+     * @param name
+     *            the name of the property
+     * @param signal
+     *            the signal to bind or <code>null</code> to unbind any existing
+     *            binding
+     * @throws com.vaadin.signals.BindingActiveException
+     *             thrown when there is already an existing binding
+     * @see #setProperty(String, String)
+     */
```
Comment (tltv) [inline flow-server/src/main/java/com/vaadin/flow/dom/Element.java:890]: Missing `return` value docs. Or change to void as with other bind* methods.

File: flow-server/src/main/java/com/vaadin/flow/dom/Element.java (hunk 4 above)
Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/dom/Element.java]: Changed to `Signal<Map<?,?>>`

File: flow-server/src/main/java/com/vaadin/flow/dom/Element.java (hunk 5 above)
Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/dom/Element.java:890]: Changed to void to comply with other bind* methods

File: flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java [hunk 6]
```diff
@@ -567,6 +651,23 @@ private Runnable doDeferredUpdateFromClient(String key, Serializable value)
...
-        return putWithDeferredChangeEvent(key, value, false);
+        PutResult putResult = null;
+        if (hasSignal(key)) {
+            SignalBinding binding = (SignalBinding) super.get(key);
+            putResult = putWithDeferredChangeEvent(key, new SignalBinding(
+                    binding.signal(), binding.registration(), value), false);
+            if (binding.signal() instanceof ValueSignal valueSignal) {
+                valueSignal.value(value);
+            } else {
+                throw new PropertyChangeDeniedException(String.format(
```
Comment (tltv) [inline flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java:661]: nit: would be nice to have a test for this case too when you actually try to bind a signal like mapped or computed signal, and it throws PropertyChangeDeniedException on change from client.

Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java:661]: True, I'll add a test for this case

Comment (tepi) [reply flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementPropertyMap.java:661]: Added

**Review by tltv — APPROVED** [review]
